### ISSUE TITLE
Add parallel rendering

### ISSUE THREAD

**Issue by thomasp85** [thomasp85/gganimate#78]
_From @thomasp85 on June 12, 2018 9:28_

Moved from #23

Hopefully something build on `future.apply` - would be nice if progress bar functionality could be preserved

_Copied from original issue: thomasp85/gganimate_fork#28_

**Comment by STJukes** [comment]
@thomasp85 Is there any update on multi-thread gganimate rendering? I am only using 1 out of 8 cores on my PC to render large animations. Thanks!

**Comment by thomasp85** [comment]
I have my priorities elsewhere at the moment but it will probably be part of the next larger release

**Comment by titaniumtroop** [comment]
I ran a `profvis::profvis` profile on the `animate` function when rendering a state-level map (attached) with 185 frames. Running the process without profiling took about 8.5 minutes on an older MacBook Pro. Here's the graphic:

![ark_covid_rolling_cases](https://user-images.githubusercontent.com/13068310/91120680-d2b96800-e65b-11ea-827e-e08acd70e479.gif)

The .Rprofvis file is pretty large (59MB), so I'll summarize instead of attaching (I will email/Dropbox if you like). I drilled down into the Data tab. About 60% of the total run time was attributable to a single `lapply` call in the `expand_layer` function in the [transition-.R](https://github.com/thomasp85/gganimate/blob/cd4e7ec05bcdaa9eebe4ddf6ac79c9b6b46d735a/R/transition-.R) file. Since the results of that function call are simply `rbind`ed together, it seems like a reasonable place to make some speed improvements with parallelization.

**Comment by titaniumtroop** [comment]
After looking at the `future.apply` documentation and noticing the availability of `future.apply::future_Map`, it looks like there's also a call to `Map` in the `expand_data` function, which then calls the `expand_layer` function mentioned above. As such, the `Map` function may be a better candidate for parallelization because it's an earlier call in the stack.

**Comment by HenrikBengtsson** [comment]
See PR #403.  It adds support for parallelizing `animate()` via futures.  The gist was to first rewrite the for-loop in `animate()` as a `mapply()` call.  Then it was only a minor tweak to make use of `future_mapply()`.

Here's a quick proof-of-concept benchmark based on `example("animate", package = "gganimate")`:

```r
library(gganimate)

anim <- ggplot(mtcars, aes(mpg, disp)) +
  transition_states(gear, transition_length = 2, state_length = 1) +
  enter_fade() +
  exit_fade()

future::plan("sequential")  ## default
t0 <- system.time(animate(anim))
print(t0)

future::plan("multiprocess", workers = 4L)
t1 <- system.time(animate(anim))
print(t1)

print(t0 / t1)
##      user    system   elapsed 
## 2.2802385 0.9121339 2.5819751
```

**Comment by jeroen** [comment]
Also note that the encoding step will be much faster and much better quality by using a proper modern video format instead of gif. For example:


```r
animate(anim, fps = 25, renderer = av_renderer("out.mkv"))
```

This uses the the av package, which has state-of-the-art multhreaded high definition encoders.

**Comment by RoyalTS** [comment]
I'm assuming this has gone stale at this point but it'd still be very nifty to have this. I, too, am working on an animated map and even on modern hardware it's taking a fair while.

**Comment by benjaminhlina** [comment]
Agreed, @thomasp85  it would be greatly appreciated if this was explored and implemented considering there's a pull request for this. I'm trying to render animated heatmaps and because of the lack of parallels processing, the time it takes to animate is long. I have been tempted to save each timestep, (a day) as an image and then use video software to render the animation by connecting the images.

**Comment by howardbaik** [comment]
@thomasp85 are there any updates to running `gganimate::animate()` in parallel?
